desktop.c: warning: ‘__builtin_strncpy’ specified bound 24 equals destination size

This is using `debuild -us -uc -rfakeroot` to build `labwc-menu-generator`

<details><summary>labwc-menu-generator_0.1.0-1_amd64.build (warning)</summary>

```
 dpkg-buildpackage -rfakefoot -us -uc -ui
dpkg-buildpackage: info: source package labwc-menu-generator
dpkg-buildpackage: info: source version 0.1.0-1
dpkg-buildpackage: info: source distribution UNRELEASED
dpkg-buildpackage: info: source changed by Mick Mars <01micko@gmail.com>
 dpkg-source --before-build .
dpkg-buildpackage: info: host architecture amd64
 debian/rules clean
dh clean
   dh_auto_clean
	make -j6 clean
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
rm -f -r 'trash directory'.* test-results
rm -f .prove
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   dh_clean
 dpkg-source -b .
dpkg-source: info: using source format '3.0 (quilt)'
dpkg-source: info: building labwc-menu-generator using existing ./labwc-menu-generator_0.1.0.orig.tar.gz
dpkg-source: info: building labwc-menu-generator in labwc-menu-generator_0.1.0-1.debian.tar.xz
dpkg-source: info: building labwc-menu-generator in labwc-menu-generator_0.1.0-1.dsc
 debian/rules binary
dh binary
   dh_update_autotools_config
   dh_autoreconf
   dh_auto_configure
   dh_auto_build
	make -j6 "INSTALL=install --strip-program=true"
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -Wdate-time -D_FORTIFY_SOURCE=2  -c -o main.o main.c
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -Wdate-time -D_FORTIFY_SOURCE=2  -c -o desktop.o desktop.c
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -Wdate-time -D_FORTIFY_SOURCE=2  -c -o ignore.o ignore.c
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -o labwc-menu-generator main.o desktop.o ignore.o -Wl,-z,relro `pkg-config --libs glib-2.0`
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   dh_auto_test
	make -j6 check
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
rm -f -r test-results
*** prove ***
t1000.t .. ok
t1001.t .. ok
All tests successful.
Files=2, Tests=2,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.02 cusr  0.00 csys =  0.03 CPU)
Result: PASS
make clean-except-prove-cache
rm -f -r 'trash directory'.* test-results
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   create-stamp debian/debhelper-build-stamp
   dh_prep
   debian/rules override_dh_auto_install
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
dh_auto_configure --buildsystem=meson
	cd obj-x86_64-linux-gnu && DEB_PYTHON_INSTALL_LAYOUT=deb LC_ALL=C.UTF-8 meson setup .. --wrap-mode=nodownload --buildtype=plain --prefix=/usr --sysconfdir=/etc --localstatedir=/var --libdir=lib/x86_64-linux-gnu -Dpython.bytecompile=-1
The Meson build system
Version: 1.4.1
Source dir: /home/mick/Packages/labwc/labwc-menu-generator-0.1.0
Build dir: /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu
Build type: native build
Project name: labwc-menu-generator
Project version: 0.1.0
C compiler for the host machine: cc (gcc 13.2.0 "cc (Debian 13.2.0-25) 13.2.0")
C linker for the host machine: cc ld.bfd 2.42
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: YES (/usr/bin/pkg-config) 1.8.1
Run-time dependency glib-2.0 found: YES 2.80.2
Program scdoc found: YES (/usr/bin/scdoc)
Build targets in project: 2

labwc-menu-generator 0.1.0

  User defined options
    buildtype         : plain
    libdir            : lib/x86_64-linux-gnu
    localstatedir     : /var
    prefix            : /usr
    sysconfdir        : /etc
    wrap_mode         : nodownload
    python.bytecompile: -1

Found ninja-1.12.1 at /usr/bin/ninja
meson compile -C obj-x86_64-linux-gnu/
ninja: Entering directory `/home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu'
[1/5] Generating data/labwc-menu-generator.1 with a custom command (wrapped by meson to capture output, to feed input)
[2/5] Compiling C object labwc-menu-generator.p/ignore.c.o
[3/5] Compiling C object labwc-menu-generator.p/main.c.o
[4/5] Compiling C object labwc-menu-generator.p/desktop.c.o
../desktop.c: In function ‘desktop_entries_create’:
../desktop.c:51:9: warning: ‘__builtin_strncpy’ specified bound 24 equals destination size [-Wstringop-truncation]
   51 |         strncpy(llcc, p, sizeof(llcc));
      |         ^
[5/5] Linking target labwc-menu-generator
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja -C /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu
meson install -C obj-x86_64-linux-gnu/ --destdir=$(pwd)/debian/labwc-menu-generator
ninja: Entering directory `/home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu'
ninja: no work to do.
Installing labwc-menu-generator to /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/debian/labwc-menu-generator/usr/bin
Installing data/labwc-menu-generator.1 to /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/debian/labwc-menu-generator/usr/share/man/man1
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   dh_installdocs
   dh_installchangelogs
   dh_installman
   dh_perl
   dh_link
   dh_strip_nondeterminism
   dh_compress
   dh_fixperms
   dh_missing
   dh_dwz -a
   dh_strip -a
   dh_makeshlibs -a
   dh_shlibdeps -a
   dh_installdeb
   dh_gencontrol
   dh_md5sums
   dh_builddeb
dpkg-deb: building package 'labwc-menu-generator' in '../labwc-menu-generator_0.1.0-1_amd64.deb'.
dpkg-deb: building package 'labwc-menu-generator-dbgsym' in '../labwc-menu-generator-dbgsym_0.1.0-1_amd64.deb'.
 dpkg-genbuildinfo -O../labwc-menu-generator_0.1.0-1_amd64.buildinfo
 dpkg-genchanges -O../labwc-menu-generator_0.1.0-1_amd64.changes
dpkg-genchanges: info: including full source code in upload
 dpkg-source --after-build .
dpkg-buildpackage: info: full upload (original source is included)
Now running lintian labwc-menu-generator_0.1.0-1_amd64.changes ...
Finished running lintian.
```
</details>

I patched it the "debian way"

<details><summary>debian/patches/000-stringop-truncation.patch</summary>

```patch
From: Mick Mars <01micko@gmail.com>
Description: set stringop-truncation patch
--- labwc-menu-generator-0.1.0.orig/desktop.c	2024-06-09 07:55:30.082028828 +1000
+++ labwc-menu-generator-0.1.0/desktop.c	2024-06-15 10:41:43.280054913 +1000
@@ -48,7 +48,7 @@
 	}
 
 	/* ll_CC */
-	strncpy(llcc, p, sizeof(llcc));
+	strncpy(llcc, p, sizeof(llcc) - 1);
 	p = strrchr(llcc, '.');
 	if (p) {
 		*p = '\0';
```

</details>

Ran `debuild` again

<details><summary>labwc-menu-generator_0.1.0-1_amd64.build (no warnings, not even `lintian` :stuck_out_tongue_closed_eyes: )</summary>

```
 dpkg-buildpackage -rfakeroot -us -uc -ui
dpkg-buildpackage: info: source package labwc-menu-generator
dpkg-buildpackage: info: source version 0.1.0-1
dpkg-buildpackage: info: source distribution UNRELEASED
dpkg-buildpackage: info: source changed by Mick Mars <01micko@gmail.com>
 dpkg-source --before-build .
dpkg-buildpackage: info: host architecture amd64
 debian/rules clean
dh clean
   dh_auto_clean
	make -j6 clean
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
rm -f -r 'trash directory'.* test-results
rm -f .prove
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   dh_clean
 dpkg-source -b .
dpkg-source: info: using source format '3.0 (quilt)'
dpkg-source: info: building labwc-menu-generator using existing ./labwc-menu-generator_0.1.0.orig.tar.gz
dpkg-source: info: using patch list from debian/patches/series
dpkg-source: info: building labwc-menu-generator in labwc-menu-generator_0.1.0-1.debian.tar.xz
dpkg-source: info: building labwc-menu-generator in labwc-menu-generator_0.1.0-1.dsc
 debian/rules binary
dh binary
   dh_update_autotools_config
   dh_autoreconf
   dh_auto_configure
   dh_auto_build
	make -j6 "INSTALL=install --strip-program=true"
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -Wdate-time -D_FORTIFY_SOURCE=2  -c -o main.o main.c
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -Wdate-time -D_FORTIFY_SOURCE=2  -c -o desktop.o desktop.c
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -Wdate-time -D_FORTIFY_SOURCE=2  -c -o ignore.o ignore.c
c99 -g -O2 -Werror=implicit-function-declaration -ffile-prefix-map=/home/mick/Packages/labwc/labwc-menu-generator-0.1.0=. -fstack-protector-strong -fstack-clash-protection -Wformat -Werror=format-security -fcf-protection -g -Wall -Os -std=c99 -Wextra `pkg-config --cflags glib-2.0` -o labwc-menu-generator main.o desktop.o ignore.o -Wl,-z,relro `pkg-config --libs glib-2.0`
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   dh_auto_test
	make -j6 check
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
rm -f -r test-results
*** prove ***
t1000.t .. ok
t1001.t .. ok
All tests successful.
Files=2, Tests=2,  0 wallclock secs ( 0.01 usr  0.00 sys +  0.02 cusr  0.00 csys =  0.03 CPU)
Result: PASS
make clean-except-prove-cache
rm -f -r 'trash directory'.* test-results
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   create-stamp debian/debhelper-build-stamp
   dh_prep
   debian/rules override_dh_auto_install
make[1]: Entering directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
dh_auto_configure --buildsystem=meson
	cd obj-x86_64-linux-gnu && DEB_PYTHON_INSTALL_LAYOUT=deb LC_ALL=C.UTF-8 meson setup .. --wrap-mode=nodownload --buildtype=plain --prefix=/usr --sysconfdir=/etc --localstatedir=/var --libdir=lib/x86_64-linux-gnu -Dpython.bytecompile=-1
The Meson build system
Version: 1.4.1
Source dir: /home/mick/Packages/labwc/labwc-menu-generator-0.1.0
Build dir: /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu
Build type: native build
Project name: labwc-menu-generator
Project version: 0.1.0
C compiler for the host machine: cc (gcc 13.2.0 "cc (Debian 13.2.0-25) 13.2.0")
C linker for the host machine: cc ld.bfd 2.42
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found pkg-config: YES (/usr/bin/pkg-config) 1.8.1
Run-time dependency glib-2.0 found: YES 2.80.2
Program scdoc found: YES (/usr/bin/scdoc)
Build targets in project: 2

labwc-menu-generator 0.1.0

  User defined options
    buildtype         : plain
    libdir            : lib/x86_64-linux-gnu
    localstatedir     : /var
    prefix            : /usr
    sysconfdir        : /etc
    wrap_mode         : nodownload
    python.bytecompile: -1

Found ninja-1.12.1 at /usr/bin/ninja
meson compile -C obj-x86_64-linux-gnu/
ninja: Entering directory `/home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu'
[1/5] Generating data/labwc-menu-generator.1 with a custom command (wrapped by meson to capture output, to feed input)
[2/5] Compiling C object labwc-menu-generator.p/ignore.c.o
[3/5] Compiling C object labwc-menu-generator.p/main.c.o
[4/5] Compiling C object labwc-menu-generator.p/desktop.c.o
[5/5] Linking target labwc-menu-generator
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja -C /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu
meson install -C obj-x86_64-linux-gnu/ --destdir=$(pwd)/debian/labwc-menu-generator
ninja: Entering directory `/home/mick/Packages/labwc/labwc-menu-generator-0.1.0/obj-x86_64-linux-gnu'
ninja: no work to do.
Installing labwc-menu-generator to /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/debian/labwc-menu-generator/usr/bin
Installing data/labwc-menu-generator.1 to /home/mick/Packages/labwc/labwc-menu-generator-0.1.0/debian/labwc-menu-generator/usr/share/man/man1
make[1]: Leaving directory '/home/mick/Packages/labwc/labwc-menu-generator-0.1.0'
   dh_installdocs
   dh_installchangelogs
   dh_installman
   dh_perl
   dh_link
   dh_strip_nondeterminism
   dh_compress
   dh_fixperms
   dh_missing
   dh_dwz -a
   dh_strip -a
   dh_makeshlibs -a
   dh_shlibdeps -a
   dh_installdeb
   dh_gencontrol
   dh_md5sums
   dh_builddeb
dpkg-deb: building package 'labwc-menu-generator' in '../labwc-menu-generator_0.1.0-1_amd64.deb'.
dpkg-deb: building package 'labwc-menu-generator-dbgsym' in '../labwc-menu-generator-dbgsym_0.1.0-1_amd64.deb'.
 dpkg-genbuildinfo -O../labwc-menu-generator_0.1.0-1_amd64.buildinfo
 dpkg-genchanges -O../labwc-menu-generator_0.1.0-1_amd64.changes
dpkg-genchanges: info: including full source code in upload
 dpkg-source --after-build .
dpkg-source: info: unapplying 000-stringop-truncation.patch
dpkg-buildpackage: info: full upload (original source is included)
Now running lintian labwc-menu-generator_0.1.0-1_amd64.changes ...
Finished running lintian.
```

</details>

The `patch` seems too simple

1. I can't reproduce this outside of the "debian way" even with `-Wstringop-truncation`, patched or original 
2. why is it only need in line line 51? Ok, I understand that `llcc` is longer than `ll` and I realise they are both allocated `[24]` and that the ` - 1` I've inserted is for `\0`. Is that a reasonable explanation?
3.  Should I offer a PR or am I being OCD about warnings in the "debian way"?

--- a/desktop.c
+++ b/desktop.c
@@ -26,6 +26,30 @@ static char name_llcc[64] = { 0 };
 static char generic_name_ll[64] = { 0 };
 static char generic_name_llcc[64] = { 0 };
 
+ /*
+  * This snippet borrowed from qemu
+  * Copyright (C) Fabrice Bellard 2006-2017 GPL-2.0
+  * https://github.com/qemu/qemu/blob/master/util/cutils.c
+  * Replaces strncpy
+  */
+void
+pstrcpy(char *buf, int buf_size, const char *str)
+{
+	int c;
+	char *q = buf;
+
+	if (buf_size <= 0)
+		return;
+
+	for(;;) {
+		c = *str++;
+		if (c == 0 || q >= buf + buf_size - 1)
+			break;
+		*q++ = c;
+	}
+	*q = '\0';
+}
+
 static void
 i18n_init(void)
 {
@@ -48,14 +72,14 @@ i18n_init(void)
 	}
 
 	/* ll_CC */
-	strncpy(llcc, p, sizeof(llcc));
+	pstrcpy(llcc, sizeof(llcc), p);
 	p = strrchr(llcc, '.');
 	if (p) {
 		*p = '\0';
 	}
 
 	/* ll */
-	strncpy(ll, llcc, sizeof(ll));
+	pstrcpy(ll, sizeof(ll), llcc);
 	p = strrchr(ll, '_');
 	if (p) {
 		*p = '\0';
@@ -303,8 +327,8 @@ process_file(char *filename, const char *path)
 		return;
 	}
 	size_t len = strlen(path);
-	strncpy(fullname, path, sizeof(fullname));
-	strncpy(fullname + len, filename, sizeof(fullname) - len);
+	pstrcpy(fullname, sizeof(fullname), path);
+	pstrcpy(fullname + len, sizeof(fullname) - len, filename);
 	FILE *fp = fopen(fullname, "r");
 	if (!fp) {
 		fprintf(stderr, "warn: could not open file %s", filename);

